File: backend/app/services/common_code_service.py
from __future__ import annotations

from datetime import datetime, timezone

from fastapi import HTTPException, status
from sqlmodel import Session, select

from app.models import AppCode, AppCodeGroup
from app.schemas.common_code import (
    ActiveCodeOption,
    CodeCreateRequest,
    CodeGroupCreateRequest,
    CodeGroupItem,
    CodeGroupUpdateRequest,
    CodeItem,
    CodeUpdateRequest,
)


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


def _group_item(group: AppCodeGroup) -> CodeGroupItem:
    return CodeGroupItem(
        id=group.id,
        code=group.code,
        name=group.name,
        description=group.description,
        is_active=group.is_active,
        sort_order=group.sort_order,
        created_at=group.created_at,
        updated_at=group.updated_at,
    )


def _code_item(code: AppCode) -> CodeItem:
    return CodeItem(
        id=code.id,
        group_id=code.group_id,
        code=code.code,
        name=code.name,
        description=code.description,
        is_active=code.is_active,
        sort_order=code.sort_order,
        extra_value1=code.extra_value1,
        extra_value2=code.extra_value2,
        created_at=code.created_at,
        updated_at=code.updated_at,
    )


def _get_group_or_404(session: Session, group_id: int) -> AppCodeGroup:
    group = session.get(AppCodeGroup, group_id)
    if group is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Code group not found.")
    return group


def _get_code_or_404(session: Session, group_id: int, code_id: int) -> AppCode:
    code = session.get(AppCode, code_id)
    if code is None or code.group_id != group_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Code not found.")
    return code


def list_code_groups(session: Session) -> list[CodeGroupItem]:
    rows = session.exec(select(AppCodeGroup).order_by(AppCodeGroup.sort_order, AppCodeGroup.id)).all()
    return [_group_item(row) for row in rows]


def create_code_group(session: Session, payload: CodeGroupCreateRequest) -> CodeGroupItem:
    code = payload.code.strip().upper()
    name = payload.name.strip()

    duplicate = session.exec(select(AppCodeGroup.id).where(AppCodeGroup.code == code)).first()
    if duplicate is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="group code already exists.")

    group = AppCodeGroup(
        code=code,
        name=name,
        description=payload.description,
        is_active=payload.is_active,
        sort_order=payload.sort_order,
        created_at=_utc_now(),
        updated_at=_utc_now(),
    )
    session.add(group)
    session.commit()
    session.refresh(group)
    return _group_item(group)


def update_code_group(session: Session, group_id: int, payload: CodeGroupUpdateRequest) -> CodeGroupItem:
    group = _get_group_or_404(session, group_id)

    if payload.code is not None:
        next_code = payload.code.strip().upper()
        duplicate = session.exec(
            select(AppCodeGroup.id).where(AppCodeGroup.code == next_code, AppCodeGroup.id != group_id)
        ).first()
        if duplicate is not None:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="group code already exists.")
        group.code = next_code

    if payload.name is not None:
        group.name = payload.name.strip()
    if payload.description is not None:
        group.description = payload.description
    if payload.is_active is not None:
        group.is_active = payload.is_active
    if payload.sort_order is not None:
        group.sort_order = payload.sort_order

    group.updated_at = _utc_now()
    session.add(group)
    session.commit()
    session.refresh(group)
    return _group_item(group)


def delete_code_group(session: Session, group_id: int) -> None:
    group = _get_group_or_404(session, group_id)
    session.delete(group)
    session.commit()


def list_codes(session: Session, group_id: int) -> list[CodeItem]:
    _get_group_or_404(session, group_id)
    rows = session.exec(
        select(AppCode).where(AppCode.group_id == group_id).order_by(AppCode.sort_order, AppCode.id)
    ).all()
    return [_code_item(row) for row in rows]


def create_code(session: Session, group_id: int, payload: CodeCreateRequest) -> CodeItem:
    _get_group_or_404(session, group_id)
    code_value = payload.code.strip().upper()
    name = payload.name.strip()

    duplicate = session.exec(
        select(AppCode.id).where(AppCode.group_id == group_id, AppCode.code == code_value)
    ).first()
    if duplicate is not None:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="code already exists in group.")

    row = AppCode(
        group_id=group_id,
        code=code_value,
        name=name,
        description=payload.description,
        is_active=payload.is_active,
        sort_order=payload.sort_order,
        extra_value1=payload.extra_value1,
        extra_value2=payload.extra_value2,
        created_at=_utc_now(),
        updated_at=_utc_now(),
    )
    session.add(row)
    session.commit()
    session.refresh(row)
    return _code_item(row)


def update_code(session: Session, group_id: int, code_id: int, payload: CodeUpdateRequest) -> CodeItem:
    row = _get_code_or_404(session, group_id, code_id)

    if payload.code is not None:
        next_code = payload.code.strip().upper()
        duplicate = session.exec(
            select(AppCode.id).where(
                AppCode.group_id == group_id,
                AppCode.code == next_code,
                AppCode.id != code_id,
            )
        ).first()
        if duplicate is not None:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="code already exists in group.")
        row.code = next_code

    if payload.name is not None:
        row.name = payload.name.strip()
    if payload.description is not None:
        row.description = payload.description
    if payload.is_active is not None:
        row.is_active = payload.is_active
    if payload.sort_order is not None:
        row.sort_order = payload.sort_order
    if payload.extra_value1 is not None:
        row.extra_value1 = payload.extra_value1
    if payload.extra_value2 is not None:
        row.extra_value2 = payload.extra_value2

    row.updated_at = _utc_now()
    session.add(row)
    session.commit()
    session.refresh(row)
    return _code_item(row)


def delete_code(session: Session, group_id: int, code_id: int) -> None:
    row = _get_code_or_404(session, group_id, code_id)
    session.delete(row)
    session.commit()


def list_active_codes(session: Session, group_code: str) -> list[ActiveCodeOption]:
    code = group_code.strip().upper()
    group = session.exec(select(AppCodeGroup).where(AppCodeGroup.code == code)).first()
    if group is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Code group not found.")

    rows = session.exec(
        select(AppCode)
        .where(AppCode.group_id == group.id, AppCode.is_active == True)
        .order_by(AppCode.sort_order, AppCode.id)
    ).all()
    return [ActiveCodeOption(code=row.code, name=row.name) for row in rows]

File: backend-spring/src/main/java/com/vibehr/commoncode/CommonCodeService.java
package com.vibehr.commoncode;

import com.vibehr.platform.error.ApiException;
import com.vibehr.platform.ids.IntegerId;
import java.time.Clock;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Locale;
import org.springframework.context.annotation.Profile;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Profile("!test")
public class CommonCodeService {

    private final CodeGroupRepository groups;
    private final CodeRepository codes;
    private final Clock clock;

    CommonCodeService(CodeGroupRepository groups, CodeRepository codes, Clock clock) {
        this.groups = groups;
        this.codes = codes;
        this.clock = clock;
    }

    @Transactional(readOnly = true)
    public PageResult<CodeGroupItem> groups(int page, int limit, boolean all, String code, String name) {
        String normalizedCode = filter(code);
        String normalizedName = filter(name);
        long total = groups.countFiltered(normalizedCode, normalizedName);
        Pageable pageable = all ? Pageable.unpaged() : PageRequest.of(page - 1, limit);
        return new PageResult<>(groups.findFiltered(normalizedCode, normalizedName, pageable).stream().map(this::toGroup).toList(), total, all ? null : page, all ? null : limit);
    }

    @Transactional
    public CodeGroupItem createGroup(CodeGroupRequest request) {
        String code = request.code().strip().toUpperCase(Locale.ROOT);
        if (groups.findByCode(code).isPresent()) throw ApiException.conflict("group code already exists.");
        return toGroup(groups.saveAndFlush(new AppCodeGroup(code, request.name().strip(), request.description(), request.isActive(), request.sortOrder(), now())));
    }

    @Transactional
    public CodeGroupItem updateGroup(long groupId, CodeGroupUpdateRequest request) {
        int id = IntegerId.required(groupId, "group_id");
        AppCodeGroup group = group(id);
        if (request.code() != null) {
            String next = request.code().strip().toUpperCase(Locale.ROOT);
            if (groups.findByCode(next).filter(found -> !found.getId().equals(id)).isPresent()) throw ApiException.conflict("group code already exists.");
        }
        group.update(request.code(), request.name(), request.description(), request.isActive(), request.sortOrder(), now());
        return toGroup(group);
    }

    @Transactional
    public void deleteGroup(long groupId) { groups.delete(group(IntegerId.required(groupId, "group_id"))); }

    @Transactional(readOnly = true)
    public PageResult<CodeItem> codes(long groupId, int page, int limit, boolean all, String code, String name) {
        int id = IntegerId.required(groupId, "group_id");
        group(id);
        String normalizedCode = filter(code);
        String normalizedName = filter(name);
        long total = codes.countFiltered(id, normalizedCode, normalizedName);
        Pageable pageable = all ? Pageable.unpaged() : PageRequest.of(page - 1, limit);
        return new PageResult<>(codes.findFiltered(id, normalizedCode, normalizedName, pageable).stream().map(this::toCode).toList(), total, all ? null : page, all ? null : limit);
    }

    @Transactional
    public CodeItem createCode(long groupId, CodeRequest request) {
        int id = IntegerId.required(groupId, "group_id");
        group(id);
        String code = request.code().strip().toUpperCase(Locale.ROOT);
        if (codes.findByGroupIdAndCode(id, code).isPresent()) throw ApiException.conflict("code already exists in group.");
        return toCode(codes.saveAndFlush(new AppCode(id, code, request.name().strip(), request.description(), request.isActive(), request.sortOrder(), request.extraValue1(), request.extraValue2(), now())));
    }

    @Transactional
    public CodeItem updateCode(long groupId, long codeId, CodeUpdateRequest request) {
        int group = IntegerId.required(groupId, "group_id");
        int id = IntegerId.required(codeId, "code_id");
        AppCode code = code(group, id);
        if (request.code() != null) {
            String next = request.code().strip().toUpperCase(Locale.ROOT);
            if (codes.findByGroupIdAndCode(group, next).filter(found -> !found.getId().equals(id)).isPresent()) throw ApiException.conflict("code already exists in group.");
        }
        code.update(request.code(), request.name(), request.description(), request.isActive(), request.sortOrder(), request.extraValue1(), request.extraValue2(), now());
        return toCode(code);
    }

    @Transactional
    public void deleteCode(long groupId, long codeId) { codes.delete(code(IntegerId.required(groupId, "group_id"), IntegerId.required(codeId, "code_id"))); }

    @Transactional(readOnly = true)
    public ActiveCodes activeCodes(String groupCode) {
        String normalized = groupCode.strip().toUpperCase(Locale.ROOT);
        AppCodeGroup group = groups.findByCode(normalized).orElseThrow(() -> ApiException.notFound("Code group not found."));
        return new ActiveCodes(normalized, codes.findByGroupIdAndActiveTrueOrderBySortOrderAscIdAsc(group.getId()).stream().map(row -> new ActiveCodeOption(row.getCode(), row.getName())).toList());
    }

    private AppCodeGroup group(int id) { return groups.findById(id).orElseThrow(() -> ApiException.notFound("Code group not found.")); }
    private AppCode code(int groupId, int codeId) {
        return codes.findById(codeId).filter(row -> row.getGroupId().equals(groupId)).orElseThrow(() -> ApiException.notFound("Code not found."));
    }
    private CodeGroupItem toGroup(AppCodeGroup row) { return new CodeGroupItem(row.getId(), row.getCode(), row.getName(), row.getDescription(), row.isActive(), row.getSortOrder(), row.getCreatedAt(), row.getUpdatedAt()); }
    private CodeItem toCode(AppCode row) { return new CodeItem(row.getId(), row.getGroupId(), row.getCode(), row.getName(), row.getDescription(), row.isActive(), row.getSortOrder(), row.getExtraValue1(), row.getExtraValue2(), row.getCreatedAt(), row.getUpdatedAt()); }
    private LocalDateTime now() { return LocalDateTime.now(clock); }
    private static String filter(String value) { return value == null || value.isEmpty() ? null : value.strip(); }

    public record PageResult<T>(List<T> rows, long totalCount, Integer page, Integer limit) { }
    public record CodeGroupItem(long id, String code, String name, String description, boolean isActive, int sortOrder, LocalDateTime createdAt, LocalDateTime updatedAt) { }
    public record CodeItem(long id, long groupId, String code, String name, String description, boolean isActive, int sortOrder, String extraValue1, String extraValue2, LocalDateTime createdAt, LocalDateTime updatedAt) { }
    public record ActiveCodeOption(String code, String name) { }
    public record ActiveCodes(String groupCode, List<ActiveCodeOption> options) { }
    public record CodeGroupRequest(String code, String name, String description, boolean isActive, int sortOrder) { }
    public record CodeGroupUpdateRequest(String code, String name, String description, Boolean isActive, Integer sortOrder) { }
    public record CodeRequest(String code, String name, String description, boolean isActive, int sortOrder, String extraValue1, String extraValue2) { }
    public record CodeUpdateRequest(String code, String name, String description, Boolean isActive, Integer sortOrder, String extraValue1, String extraValue2) { }
}
